### PR TITLE
Add a TSDB rate aggregation

### PR DESCRIPTION
**Pull request by romseygeek**
This commit adds an extension to the existing `rate` aggregation that, when
applied to a field defined as a metric `counter`, will calculate rates that take
into account counter resets.

### CODE REVIEW

**Comment by romseygeek** [comment]
We still need to add rate units for an MVP and should discuss how we handle the gaps over resets.

**Comment by iverase** [comment]
If I am reading correct, this implementation assumes that the rate aggregation over a counter metric will be always be used on a time series context:

`time_series agg -> rate agg`

I guess it would be possible for a user to use a rate aggregation over a counter metric on a non tsdb context:

`terms agg- >  rate agg`

 things would break badly in this case?

**Comment by iverase** [comment]
My understanding is that in order for aggregations to work on tsdb mode, they need to have a time_series aggregation as parent because it overrides the following method:

```
   @Override
    public boolean isInSortOrderExecutionRequired() {
        return true;
    }
```
My point is that the index can be a time series index but it does not necessarily needs to run all the time in sort order?

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:439]: @not-napoleon and I were just talking about this sort of thing. It makes sense here but I wonder if @not-napoleon has a different idea?

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRate.java
Comment (nik9000) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRate.java:76]: No.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java
Comment (not-napoleon) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:1513]: I think the extra constructor is too much.  There's already a `with...` method, just chain that onto the constructor calls.  I'm trying to avoid having a constructor for every set of parameters here.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java
Comment (not-napoleon) [reply test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:439]: Yeah, my other idea is "let's actually finish cleaning up aggregations tests."  But that's weeks of work with very little gain, and there's no sense in blocking this PR on that work.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java
Comment (not-napoleon) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java:205]: I don't love this.  I'd rather that the `IndexFieldData` object that this field eventually generates would report its `ValuesSourceType` as `TimeSeriesValuesSourceType.COUNTER`.  That does mean that we'd have to explicitly add `COUNTER` as a supported values source type for other aggregations, but that's not too hard.

My thinking is that we will eventually want to specialize other aggregations for time series, and I don't want to have a custom `resolveConfig` each time we want to do that.  Indeed, that kind of goes against the entire point of the `ValuesSourceRegistry`.

Having said that, I'm not sure how much work it would be to move this information into `IndexFieldData`.  Happy to discuss further if there are concerns around that.

**Review by not-napoleon — COMMENTED** [review]
I reviewed this from how it integrates with the rest of the aggregations framework.  I did not dig into the details of the algorithm for correctness (although I can if that needs another pair of eyes).

I think we need to have a bigger conversation about how time series values sources are going to work.  I left my initial thoughts inline here, but we probably should discuss it in a higher bandwidth mode.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsPlugin.java
Comment (martijnvg) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsPlugin.java:127]: Should this for now only be registered when time series mode is enabled? (by adding a `IndexSettings.isTimeSeriesModeEnabled()` if statement)

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java
Comment (romseygeek) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java:205]: An additional wrinkle is that we ideally want to only select this agg implementation if `rate` is nested under a `tsid` agg, and @iverase and I thought this would be the best place to do that - extend `resolveConfig` to take the parent agg builder and then iterate through parents to see if there's a `tsid` agg somewhere in the tree.

File: x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java
Comment (nik9000) [inline x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java:59]: I think it'd probably be useful to implement this to test random reduction. It can be a bit silly feeling in some aggs but sometimes it shows up really interesting stuff.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/TimeSeriesRateAggregator.java
Comment (martijnvg) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/TimeSeriesRateAggregator.java:92]: If `leafValues` is `null` then we can return `NO_OP_COLLECTOR`?

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/TimeSeriesRateAggregator.java
Comment (romseygeek) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/TimeSeriesRateAggregator.java:51]: If we keep them in then the constructor can be used as an implementation of `RateAggregatorSupplier`.  I'll add a comment explaining why they're there.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Review by martijnvg — APPROVED** [review]
It does seem that some of Nik's comments were not addressed? LGTM otherwise.

File: x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java
Comment (martijnvg) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java:69]: So in order to detect resets between `InternalTrackingRate` instances, these instances will need to be merged in order. I think most of the time no reduce will happen for `rate` agg. This because it needs to be wrapped in a `time_series` agg. Only for time series agg buckets with same key (same tsid) will need to do a reduce. So when the range of the parent bucket of time series bucket crosses backing index boundary then resets may not be detected.

But in any case the `InternalResetTrackingRate` instance need to be merged in order and with partial reduce I think we can't guarantee that? Maybe this `supportsOutOfOrderReduce` should be a flag in `InternalAggregation`? This can then be set by `InternalResetTrackingRate` and then no partial reduce will be triggered?

Comment (romseygeek) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java:69]: This agg will work with partial reduce *as long as the reduction is in-order* - in other words, we can reduce buckets from time 1 - 10, and then further reduce that combined bucket with buckets 11 - 20, but we can't reduce buckets 1, 2, 4, 7 and 9, and then try and merge in bucket 3.  

As I understand it, partial reduction is used by CCS and by async search; async search will always reduce in order so that it can provide meaningful partial results, so we have no problem there.  And TSDB doesn't support tsids across clusters, so the CCS situation should also not apply here.

The `supportsOutOfOrderReduce` flag doesn't stop us testing partial reductions, it stops us selecting a random set of internal buckets to reduce.  When this flag is set to `false`, partial reductions will always be in-order.

Comment (martijnvg) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java:70]: > partial reduction is used by CCS and by async search; async search will always reduce in order so that it can provide meaningful partial results, so we have no problem there. 

👍 

(I think partial reduce also happens during regular search depending `SearchRequest#batchedReduceSize`)

> And TSDB doesn't support tsids across clusters, so the CCS situation should also not apply here.

True, a tsid across data streams or clusters is impossible.

> The supportsOutOfOrderReduce flag doesn't stop us testing partial reductions, it stops us selecting a random set of internal buckets to reduce. When this flag is set to false, partial reductions will always be in-order.

Agreed, then my previous comment about having a flag `InternalAggregation` makes no sense and can be ignored.

Comment (nik9000) [reply x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java:70]: I hadn't realized that async search would reduce in order. Ok.

**Review by nik9000 — APPROVED** [review]

**Review by nik9000 — APPROVED** [review]

**Pull request merged**
